src/renderBlocks.ts

```diff
@@ -179,6 +179,8 @@ function render(bodyElements: BlockElement[], imageSalt: string, ads = true): Re
 
 function componentFromHtml(html: string): ReactNode[] {
     const fragment = JSDOM.fragment(transform(html))
+    Array.from(fragment.children)
+        .forEach((element, index) => element.setAttribute('key', `${index}`))
```

Comment:
I've been trying to find another way to achieve the same thing, but I can't come up with something without having to fall into hashing the content of a block to identify it...

Somehow it feels wrong to simply assign a number with each component. Any thoughts on that?

Reply:
Is it the concept of using a number in general that feels wrong, or is it manipulating the DOM to add an attribute?

I think applying an index to *React components* is a fairly straightforward method when working with arrays, because you typically already have it if you're mapping/reducing, Assigning a randomly generated value (e.g. using uuid) is another technique, although I'm not keen on that because it breaks referential transparency.

This is solved in the new renderer [like this](https://github.com/guardian/apps-rendering/blob/f8b2c7033138363510f7255ad4a6798a749f2239/src/block.ts#L189), which is possible because the code is structured differently.

Reply:
Having done some research I'm pretty sure I'm wrong about this... You're right @alexduf, the [react docs](https://reactjs.org/docs/lists-and-keys.html#keys) explain that using index as the key is a problem if the list is likely to change.

In practice, I don't think we'd see any problems from this because it's a single render. However, on the client this may result in bugs due to possible re-renders (particularly on the liveblog). As we're sharing code between the client and server, perhaps it would be best to eliminate this pattern entirely. @webb04 @alexduf what do you think?

Reply:
I think we can decide that this is a relevant issue only on the client.

Since the server-side code and client-side code are built separately, perhaps we can disable this warning for the server build only?

Reply:
Doesn't seem an easy way to turn off warnings, unless we use production React which will remove a lot more warnings.

Reply:
`componentFromHtml` should never make it's way to the client.

Reply:
I guess it makes sense then